lib/EPub.php

```diff
@@ -0,0 +1,843 @@
+        $nodes = $this->xpath->query('//opf:metadata/opf:meta[@name="cover"]');
+        if (!$nodes->length) {
+            return $this->no_cover();
+        }
+        $coverid = (String) $nodes->item(0)->attr('opf:content');
+        if (!$coverid) {
+            return $this->no_cover();
+        }
+
+        $nodes = $this->xpath->query('//opf:manifest/opf:item[@id="' . $coverid . '"]');
+        if (!$nodes->length) {
+            return $this->no_cover();
+        }
+        $mime = $nodes->item(0)->attr('opf:media-type');
+        $path = $nodes->item(0)->attr('opf:href');
+        $path = dirname('/' . $this->meta) . '/' . $path; // image path is relative to meta file
+        $path = ltrim($path, '/');
+
+        $zip = new ZipArchive();
+        if (!@$zip->open($this->file)) {
+            throw new Exception('Failed to read epub file');
+        }
+        $data = $zip->getFromName($path);
+
+        return array(
+            'mime'  => $mime,
+            'data'  => $data,
+            'found' => $path
+        );
+    }
+
+    public function getCoverItem()
+    {
+        $nodes = $this->xpath->query('//opf:metadata/opf:meta[@name="cover"]');
+        if (!$nodes->length) {
+            return NULL;
+        }
+
+        $coverid = (String)$nodes->item(0)->attr('opf:content');
+        if (!$coverid) {
+            return NULL;
+        }
+
+        $nodes = $this->xpath->query('//opf:manifest/opf:item[@id="' . $coverid . '"]');
+        if (!$nodes->length) {
+            return NULL;
+        }
+
+        return $nodes->item(0);
+    }
+
+    public function Combine($a, $b)
+    {
+        $isAbsolute = false;
+        if ($a[0] == '/') {
+            $isAbsolute = true;
+        }
+
+        if ($b[0] == '/') {
+            throw new InvalidArgumentException('Second path part must not start with ' . $m_Separator);
+        }
+
+        $splittedA = preg_split('#/#', $a);
+        $splittedB = preg_split('#/#', $b);
+
+        $pathParts = array();
+        $mergedPath = array_merge($splittedA, $splittedB);
+
+        foreach ($mergedPath as $item) {
+            if ($item == null || $item == '' || $item == '.') {
+                continue;
+            }
+
+            if ($item == '..') {
+                array_pop($pathParts);
+                continue;
+            }
+
+            array_push($pathParts, $item);
+        }
+
+        $path = implode('/', $pathParts);
+        if ($isAbsolute) {
+            return('/' . $path);
+        } else {
+            return($path);
+        }
+    }
+
+    private function getFullPath($file, $context = NULL)
+    {
+        $path = dirname('/' . $this->meta) . '/' . $file;
+        $path = ltrim($path, '\\');
+        $path = ltrim($path, '/');
+        if (!empty($context)) {
+            $path = $this->combine(dirname($path), $context);
+        }
+        //error_log ("FullPath : $path ($file / $context)");
+        return $path;
+    }
+
+    public function updateForKepub()
```

Comment:
Kobo don't use the standard EPUB for their own books coming from their store. 

Instead they use their own format Kepub (You can have more information [here](view-source:http://wiki.mobileread.com/wiki/Kepub#Kepub_compared_to_ePub)). Simply renaming `test.epub` to `test.epub.kepub` before sideloading it to a Kobo eReader will make it more or less handled as a Kepub.

In this function I only added a mandatory key allowing the cover to be correctly detected by Kobo eReader. I initially wanted to alter the HTML to add all the span but Kobo changed the way they handled sideloaded Kepub.